### PR TITLE
Alive blog post

### PR DESCRIPTION
**Pull request by stp59**
Here is a draft of our blog post with discussion questions for next monday!

### CODE REVIEW

File: content/blog/2020-10-04-alive/index.md [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+
+## Contributions
+
+This paper introduces Alive, a Domain-Specific Language designed for conveniently verifying LLVM optimizations. The Alive language provides a framework for encoding LLVM transformations of compiled code. The implementation uses a transformation encoding to generate preconditions and correctness conditions expressed in terms of predicates, which may then be automatically verified by an SMT solver. The result is a verified LLVM optimization in Alive, which also provides a tool for extracting the optimization into C++ code. As an additional feature, the language supports built-in predicates which are implemented by dataflow analyses.
```
Comment (sampsyo) [inline content/blog/2020-10-04-alive/index.md]: Probably don’t capitalize DSL. (You can italicize if you want to introduce the term.)

File: content/blog/2020-10-04-alive/index.md [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+[[extra.authors]]
+name = "Samwise Parkinson"
++++
+
+Verification of "things computers do" come in many flavors. In the context of compilers, the [CompCert](compcert) project is one approach where compilers are proven to never compile incorrectly. However, this fully verified formal approach comes at the cost of not supporting all language constructs and cannot optimize as well as current compilers. [Random](diff-testing) [testing](comp-validation) is another approach where random test inputs are generated in an attempt to find [bugs](find-bugs-c). These approaches can be highly effective both in terms of finding bugs and ease of use but have no correctness guarantees.
+
+This week's [paper](paper) attempts to combine the usability aspect from testing with the guarantees aspect of formal approaches. The proposed tool, Alive, formalizes our notions of compiler correctness to treat a source program as the ground truth and verify specific optimizations, namely peephole, that transform these programs. To be practical, Alive introduces a new domain-specific language (DSL) that is meant to resemble LLVM code which makes it easier to express what is meant to be verified.
```
Comment (sampsyo) [inline content/blog/2020-10-04-alive/index.md]: Maybe explain what a peephole optimization is?

File: content/blog/2020-10-04-alive/index.md [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+
+## Contributions
+
+This paper introduces Alive, a Domain-Specific Language designed for conveniently verifying LLVM optimizations. The Alive language provides a framework for encoding LLVM transformations of compiled code. The implementation uses a transformation encoding to generate preconditions and correctness conditions expressed in terms of predicates, which may then be automatically verified by an SMT solver. The result is a verified LLVM optimization in Alive, which also provides a tool for extracting the optimization into C++ code. As an additional feature, the language supports built-in predicates which are implemented by dataflow analyses.
+
+The main difficulty with verifying LLVM opmtimizations is that the source code may have undefined behaviors. For this reason, many of the interesting design decisions for Alive are related to handling undefined behavior. In principle, Alive is intended to check that optimizations are *refining*, that is, that the set of behaviors of the target code is included in the set of behaviors of the source.
```
Comment (sampsyo) [inline content/blog/2020-10-04-alive/index.md]: “Undefined behavior” is singular.

File: content/blog/2020-10-04-alive/index.md [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+
+## Detailed Summary
+
+At a high-level, the process of verifying a peephole optimization in Alive is as follows:
+
+1. Specify the transformation in Alive's domain-specific language (DSL).
+2. Alive discharges SMT formulas that encode the semantics of LLVM operations and datatypes.
+3. The formulas from step 2 are used as verification conditions and passed to an SMT solver.
+3. SMT solver either verifies the optimization as valid or returns a counterexample showing why the optimization is wrong.
+4. (Optional step) If optimization is correct, Alive synthesizes C++ code that can then generate the relevant optimization to optimize LLVM code.
```
Comment (sampsyo) [inline content/blog/2020-10-04-alive/index.md]:
```suggestion
4. (Optional step) If the optimization is correct, Alive synthesizes C++ code that can then generate the relevant optimization to optimize LLVM code.
```

Comment (sampsyo) [reply content/blog/2020-10-04-alive/index.md]: generate -> implement?

File: content/blog/2020-10-04-alive/index.md [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+Pre: C1 & C2 == 0 && MaskedValueIsZero(%V, ∼C1)
+%t0 = or   %B,  %V
+%t1 = and %t0,  C1
+%t2 = and  %B,  C2
+%R  = or  %t1, %t2
+=>
+%R  = and %t0, (C1 | C2)
+```
+
+Here the left hand side is everything before the `=>` and the right hand side is everything after. The DSL abstracts LLVN semantics: anything beginning with a capital C is a constant, the pattern %t[num] are temporary registers and the lack of data-types indicates that this transformation is valid for all LLVM datatypes. The keyword `Pre:` allows the user of the DSL to specify a pre-condition which abstracts the results that an LLVM compiler may infer from dataflow analyses before applying a transformation. These predicates are hard-coded into Alive.
```
Comment (sampsyo) [inline content/blog/2020-10-04-alive/index.md]: “LLVN”

File: content/blog/2020-10-04-alive/index.md [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+
+### Correctness of Optimizations
+
+The goal of Alive is to prove that a target optimization refines the behavior of a source program under the presence of undefined behavior.  Undefined behavior is the result of the assumptions a compiler makes about certain instructions of a program. When an instruction with undefined behavior is used then the compiler may replace it with an arbitrary sequence of instructions or assume that such undefined behavior never occurs. For example, in `C` the instruction `x + 1 > x` can be replaced with `true` (i.e. any value `!= 0`) as signed overflow is undefined behavior. The guiding principle in Alive, and really for any other compiler, is:
```
Comment (sampsyo) [inline content/blog/2020-10-04-alive/index.md]: “i.e.” always needs a comma: https://capra.cs.cornell.edu/styleguide/#egie

File: content/blog/2020-10-04-alive/index.md [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+
+*A compiler should never introduce new behavior when there is no undefined behavior but can produce new results in the presence of undefined behavior.*
+
+A nice summary of Alive and a discussion on undefined behavior can be foundin this [blog post](alive-blog) by the authors of the paper.
+
+#### undef and poison
+
+In LLVM there are three types of undefined behavior. The first arises from the keyword `undef`. `undef` in LLVM represents any value (given a specific width) each time it is read from. For example, in the program
```
Comment (sampsyo) [inline content/blog/2020-10-04-alive/index.md]: I recommend adding a : at the end (and in similar places below).

File: content/blog/2020-10-04-alive/index.md [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+%z = xor i8 undef, undef
+```
+
+`%z` can be any value in the range `{0, ..., 255}` as we are taking the xor of any two 8-bit values. A more interesting example is
+
+```
+%z = or i8 1, undef
+```
+
+where the value of `%z` becomes any odd integer in the range `{0, ..., 255}`. `undef` allows the compiler to aggresively optimize a program as LLVM [makes the assumption](undef-val) that whenever an `undef` is seen "the program is well-defined no matter what value is used."
```
Comment (sampsyo) [inline content/blog/2020-10-04-alive/index.md]: Maybe missing a verb?

File: content/blog/2020-10-04-alive/index.md [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+>
+> (2) the target is poison-free when the source is poison-free, and
+>
+> (3) the source and the target produce the same result when the source is defined and poison-free.
+
+The SMT formulas generated by Alive encode all possible types for the instructions in the source and target programs and also encode `undef`, `poison` and instruction attributes to model undefined behavior. Three main formulas corresponding to the correctness criterias above are generated and passed to the SMT solver for verification.
+
+### Example
+
+The optimization of replacing `(X << C1) / C2 ` with `X / (C2 >> C1)` whenever `C2 is a multiple of C1` can be written in Alive as follows
```
Comment (sampsyo) [inline content/blog/2020-10-04-alive/index.md]: No need for backticks around that English condition.

File: content/blog/2020-10-04-alive/index.md [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+
+The authors of Alive opened a [bug report](pr21245) and a fix was accepted where the pre-condition of the optimization was strengthened to 
+
+```
+Pre: C2 % (1<<C1) == 0 && C1 != width(C1)-1
+```
+
+## Alive's Impact
+
+At the time of publication, Alive's authors manually translated 334 LLVM peephole optimizations (from [InstCombine](instcombine)) to Alive out of a possible 1028 instructions, meaning 694 were not processed for verification. Out of the translated optimizations, the authors found 8 bugs where the most uncommon bug was due to the introduction of undefined behavior. The authors state that most of the time Alive runs in a few seconds while for instructions with multiplication and division it "can take several hours or longer to verify the larger bit-widths" as most SMT solvers struggle with such inputs. The remaining optimizations could not be translated as they include instructions that were not supported by Alive at the time (such as floating-point operations).
```
Comment (sampsyo) [inline content/blog/2020-10-04-alive/index.md]: “instructions” -> optimizations?

File: content/blog/2020-10-04-alive/index.md [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+
+While we couldn't find any documentation on the LLVM website showing us that LLVM developers were using Alive, it is still nice that the authors were able to get their tool used and tested before publication. This comment suggests that Alive is not burdensome to use and verify optimizations.
+
+### Alive2
+
+Following this work, Alive's authors have improved on Alive in multiple ways including [floating-point support](alive-fp), a precondition [inference](alive-infer) tool for optimizations and the formalization of Alive, called [AliveInLean](alive-lean), in the [Lean](lean) theorem prover (this work assumes that "proof obligations are correctly discharged by an SMT solver"). Most recently, the authors of Alive have switched Alive to [maintenance mode](alive-git) and introduced a newer version of Alive, called [Alive2](alive2-git). A nice introduction to Alive2 can be found in a series of blog posts ([1](alive2-blog1), [2](alive2-blog2), [3](alive2-blog3)). Alive2 supports regular LLVM code in addition to the DSL of Alive along with bidirectional refinement (the `=>` of Alive can be replaced with `<=>`). So far, Alive2 has found 58 total [bugs](alive2-bugs) in LLVM & [Z3](z3).
```
Comment (sampsyo) [inline content/blog/2020-10-04-alive/index.md]: This is not the right Markdown syntax for links... maybe try a Markdown previewer of some sort? For example, right here on GitHub:
https://github.com/sampsyo/cs6120/blob/bb5a40edd81981264d650919b85cd6b2df2c146b/content/blog/2020-10-04-alive/index.md

File: content/blog/2020-10-04-alive/index.md [hunk 12]
```diff
@@ -0,0 +1,227 @@
...
+
+### Contributions to LLVM (Detour)
+
+Some of the authors of this paper (everyone except Santosh Nagarakatte) have pushed for the [removal of `undef`](remove-undef-llvm) from LLVM and introduce a new construct they call `freeze`. `freeze` is a compiler construct that stops the propagation of `poison` and instead forces the instruction to return an arbitrary value (similar to `undef`.) `freeze `was [added](freeze-twitter) to [LLVM](freeze-llvm) but `undef` was not removed.
```
Comment (sampsyo) [inline content/blog/2020-10-04-alive/index.md]:
```suggestion
Some of the authors of this paper (everyone except Santosh Nagarakatte) have pushed for the [removal of `undef`](remove-undef-llvm) from LLVM and introduce a new construct they call `freeze`. `freeze` is a compiler construct that stops the propagation of `poison` and instead forces the instruction to return an arbitrary value (similar to `undef`.) `freeze` was [added](freeze-twitter) to [LLVM](freeze-llvm) but `undef` was not removed.
```

File: content/blog/2020-10-04-alive/index.md [hunk 13]
```diff
@@ -0,0 +1,227 @@
...
+
+In general, computer science education considers undefined behavior a harmful concept. However, we also contend with the fact that undefined behavior might be required for aggressive optimizations and speculative execution. Alive seems to take a third route as articulated in Regher's blog post and documented in the LLVM documentation: undefined behavior is ok as long as it "refines' ' the original (source) program in some way. Here are two quotes from the aforementioned article.
```
Comment (sampsyo) [inline content/blog/2020-10-04-alive/index.md]: Quotation marks need fixing?

**Review by sampsyo — CHANGES_REQUESTED** [review]
Great work, y’all! Please don’t forget to announce your draft on Zulip, even before the post is finalized.

File: content/blog/2020-10-04-alive/index.md (hunk 1 above)
Comment (goki0607) [reply content/blog/2020-10-04-alive/index.md]: Fixed.

File: content/blog/2020-10-04-alive/index.md (hunk 3 above)
Comment (goki0607) [reply content/blog/2020-10-04-alive/index.md]: Fixed.

File: content/blog/2020-10-04-alive/index.md (hunk 5 above)
Comment (goki0607) [reply content/blog/2020-10-04-alive/index.md]: Ooops. Fixed.

File: content/blog/2020-10-04-alive/index.md (hunk 6 above)
Comment (goki0607) [reply content/blog/2020-10-04-alive/index.md]: Fixed.

I might be wrong about this but British English style guides do not require this.

File: content/blog/2020-10-04-alive/index.md (hunk 9 above)
Comment (goki0607) [reply content/blog/2020-10-04-alive/index.md]: Fixed.

File: content/blog/2020-10-04-alive/index.md (hunk 10 above)
Comment (goki0607) [reply content/blog/2020-10-04-alive/index.md]: Fixed.

File: content/blog/2020-10-04-alive/index.md (hunk 11 above)
Comment (goki0607) [reply content/blog/2020-10-04-alive/index.md]: Sorry about this! It's been fixed.

File: content/blog/2020-10-04-alive/index.md (hunk 13 above)
Comment (goki0607) [reply content/blog/2020-10-04-alive/index.md]: Fixed.

File: content/blog/2020-10-04-alive/index.md (hunk 8 above)
Comment (goki0607) [reply content/blog/2020-10-04-alive/index.md]: Fixed.

File: content/blog/2020-10-04-alive/index.md (hunk 7 above)
Comment (goki0607) [reply content/blog/2020-10-04-alive/index.md]: Done.

File: content/blog/2020-10-04-alive/index.md (hunk 2 above)
Comment (goki0607) [reply content/blog/2020-10-04-alive/index.md]: Fixed.

**Review by sampsyo — APPROVED** [review]
Looks great! Merging now.

**Pull request merged**